Support server-side execution

davidbrochart opened the pull request:
## References

Closes #2833.
Closes #12867.

Needs:
- ~~https://github.com/jupyterlab/jupyter-collaboration/pull/217~~
- ~~https://github.com/jupyter-server/jupyter_ydoc/pull/198~~
- ~~https://github.com/jupyter-server/jupyter_ydoc/pull/201~~
- `jupyter-collaboration >=2.0`
- `jupyverse >=0.4.0`

Currently, the only Jupyter server that supports this execution mode is [Jupyverse](https://github.com/jupyter-server/jupyverse).

## Code changes

A new `server-side-execution` flag allows to execute a notebook server-side (false by default). In this execution mode, the frontend doesn't use the kernel protocol over WebSocket, but interacts with the notebook only through its shared model. Thus, this mode needs [jupyter-collaboration](https://github.com/jupyterlab/jupyter-collaboration).

## User-facing changes

The user can close their JupyterLab browser window, reopen it later and recover the notebook state, including (live) cell outputs.

## Backwards-incompatible changes

None.

matthewwiese commented:
This is really great David! So cool to follow your progress on this. 🙂

davidbrochart commented:
Here is a demo:

https://github.com/jupyterlab/jupyterlab/assets/4711805/c65b5253-da77-4203-9223-b9097e6688c6

krassowski commented:
> server-side-execution

While I see why you named it "server-side-execution", I wonder if naming it "server-side-composition" could potentially avoid confusion with the execution which happens in kernels? It feels like the fact that execution is triggered from server (well, really proxied) is an implementation detail and what users care about is where the execution results are composed into the document.

davidbrochart commented:
I agree that "server-side-execution" is not great, because kernels always execute code in the server. The main difference here is that:
- execution is triggered by a higher-level protocol (an HTTP POST instead of the frontend talking to the kernel through the kernel protocol over WebSocket), and the backend is responsible for talking to the kernel through the kernel protocol over ZMQ.
- cell outputs are not populated by the frontend, but by the backend, and the frontend sees the output changes reflected in their shared model.

But I'm not sure about "server-side-composition" either. Maybe "server-side-outputs"?

jtpio commented:
> but interacts with the notebook only through its shared model. Thus, this mode needs [jupyter-collaboration](https://github.com/jupyterlab/jupyter-collaboration).

Curious if this could be implemented in `jupyter-collaboration` directly then? Just trying to see if this implicit dependency could create issues in the future since `jupyterlab` and `jupyter-collaboration` evolve at a different pace.

davidbrochart commented:
The execution command is done through an HTTP POST instead of using the kernel protocol, so there are changes to be done in JupyterLab:
https://github.com/jupyterlab/jupyterlab/blob/21cc49a32669a998d2f73f5b0ed3f4ea9793a40e/packages/notebook/src/actions.tsx#L2480-L2500
But I'm not sure it's the right place.

davidbrochart commented:
Here is a way to try it out:

```bash
mkdir jupyterlab
cd jupyterlab
git clone https://github.com/davidbrochart/jupyterlab.git
git clone https://github.com/davidbrochart/jupyverse.git
pixi init
pixi add nodejs pip "python<3.12"
pixi shell

cd jupyterlab
git checkout server-side-execution
pip install -e .cd ..

cd jupyverse
git checkout ywidget-subdoc
pip install -e jupyverse_api
pip install -e plugins/noauth
pip install -e plugins/contents
pip install -e plugins/frontend
pip install -e plugins/lab
pip install -e plugins/jupyterlab
pip install -e plugins/kernels
pip install -e plugins/yjs
pip install -e plugins/nbconvert
pip install -e plugins/terminals
pip install -e .
cd ..

pip install ypywidgets==0.6.3
pip install jupyter-collaboration

mkdir tmp
cd tmp
jupyverse --set jupyterlab.dev_mode=true --set jupyterlab.server_side_execution=true --set kernels.require_yjs=true
```

Open a browser window at http://127.0.0.1:8000, it should open JupyterLab.
Create a notebook with the following cell and run it:
```py
from time import sleep

for i in range(1000000):
    print(i)
    sleep(2)
```
Close your browser window and open a new one at http://127.0.0.1:8000. Open the notebook, you should see the live outputs.

jtpio commented:
Also still curious if this could be implemented in `jupyter-collaboration` directly (https://github.com/jupyterlab/jupyterlab/pull/15448#issuecomment-1833511805)? If it's not, maybe we can provide an extension point to do it as an alternative?

jtpio commented:
> Where would this extension point be?

Currently there is no such extension point, as the notebook execution logic seems to be living in a `Private` namespace which makes it difficult to customize indeed. But ideally we could imagine having a plugin that would allow for using a different execution mode (WebSocket or HTTP POST), that would be made available to the notebook actions.

davidbrochart commented:
Yes I was looking for something like that, but eventually took a shortcut. What I like in the current solution is that it's simple and it also doesn't drop the connection to the kernel through the kernel protocol over WebSocket, which means that the kernel busy indicator still works, and widgets should also still work (although they won't be recovered from the notebook state).
I see this PR as an incremental step towards completely getting rid of the kernel protocol over WebSocket in the frontend, which will require more work.

davidbrochart commented:
I released Jupyverse v0.4.0, so testing this PR is now simpler:
```bash
mkdir jupyterlab
cd jupyterlab
git clone https://github.com/davidbrochart/jupyterlab.git
pixi init
pixi add nodejs pip "python<3.12"
pixi shell

cd jupyterlab
git checkout server-side-execution
pip install -e .

pip install "jupyverse[jupyterlab,noauth]"
pip install "ypywidgets>=0.6.4"
pip install "jupyter-collaboration>=2.0.0"

mkdir tmp
cd tmp
jupyverse --set jupyterlab.dev_mode=true --set jupyterlab.server_side_execution=true --set kernels.require_yjs=true
```
I changed the [top comment](https://github.com/jupyterlab/jupyterlab/pull/15448#issue-2014190659) and updated what is needed for this feature to be enabled (mainly, `jupyverse` and `jupyter-collaboration`).
Since this PR externalizes almost everything in the backend, except for the HTTP POST to execute a cell, which is under a flag, I don't see any reason to delay merging, what do you think?

krassowski commented:
> I don't see any reason to delay merging, what do you think?

Main still tracks 4.1, so my suggestion would be to merge as soon as this is reviewed and 4.2 branch out (unless there is a consensus to make an exception and add it in 4.1 but it is _really late_ for this).

Wh1isper commented:
I prefer to name this feature as `server-side-save/saving`,  

Previously we performed the save operation in the browser, now we perform the save operation on the server side and the browser just receives the push from the server side (via collaboration).

It's very pleased to see `Jupyverse` offering more features, which will motivate us to evolve towards `Jupyverse`!

davidbrochart commented:
Now that JupyterLab v4.1.0 is out, should we go ahead with this PR?

echarles commented:
I see needed definitions and clarifications on the proposed features. Looking at the changes, only cellid and docid are sent, which is fine for server-side model. There should be a way to continue supporting code execution via REST with client-side model.

davidbrochart commented:
Yes, in order to execute a cell of a notebook, the notebook document ID and the notebook cell ID are needed.
Support for client-side execution is still there, since the new server-side execution mode is under a flag.

jtpio commented:
> But ideally we could imagine having a plugin that would allow for using a different execution mode (WebSocket or HTTP POST), that would be made available to the notebook actions.

FYI @fcollonval just opened a PR to allow for something like this: https://github.com/jupyterlab/jupyterlab/pull/15830. Thanks Fred!

EduardDurech commented:
Does anything still need to be done to merge this? I can give some time as this is critical and still the top 👍 issue after many years

In the meantime until the next release, can we just build locally and run this to get the functionality?

davidbrochart commented:
I'm sorry this is taking so long, I'd love it to be released in JupyterLab 4.2.
But yes building JupyterLab with this PR is all that is needed to try it out, since Jupyverse has been released with the feature.

jtpio commented:
https://github.com/jupyterlab/jupyterlab/pull/15830 is likely the way forward with this, with `jupyter-collaboration` providing its own `ICellExecutor`.

davidbrochart commented:
Sure, #15830 is the right solution long-term, but maybe we could live with the shortcut provided by this PR in the meantime?

jtpio commented:
#15830 is almost ready and also planned for 4.2.0. So I'm not sure there is a need to merge this PR in the meantime.

Once https://github.com/jupyterlab/jupyterlab/pull/15830 is merged and available in a 4.2 pre-release, the logic in this PR can be implemented in `jupyter-collaboration` directly instead.

krassowski commented:
@davidbrochart FYI https://github.com/jupyterlab/jupyterlab/pull/15830 is now merged.

davidbrochart commented:
Thanks, I'll update this PR.

krassowski commented:
@davidbrochart is the plan to add a new implementation of `ICellExecutor` as a new plugin in this PR or to add it in `jupyter-collaboration` for time being and possibly bring the non-RTC bits of jupyter-collaboration into the core in the next release, say 4.3 (we are already past feature freeze on 4.2)?

davidbrochart commented:
You're right, I think a new implementation of `ICellExecutor` as a new plugin should be done in jupyter-collaboration, as this execution mode doesn't make sense in non-RTC mode.

davidbrochart commented:
Closing in favor of https://github.com/jupyterlab/jupyter-collaboration/pull/279.
